From 5430af6a309f486494d46fdeec923f7de0860e15 Mon Sep 17 00:00:00 2001
From: Mark Punzalan <markpun@google.com>
Date: Thu, 11 Sep 2025 20:10:11 -0700
Subject: [PATCH] Ensure there is at least one locale in the Configuration
 object

When there is no intersection between the system locales and the
app-supported locales defined in the LocaleConfig, the locales in the
Configuration object are set to an empty LocaleList, which can trigger a
crash (NullPointerException) in the app and possibly in system_server or
systemui in some cases.

This fixes that issue by falling back in that case to the previous
behavior (before the improved resource multi-locale support) of
selecting the best matching locale available in the app (if any), or the
first system locale.

Additionally:
- Renamed `calcConfigChanges` to `applyConfigChanges` because it applies the new configuration in addition to calculating the changes.
- Added more comments to explain the locale selection logic.

Bug: 400163083
Test: Manually on Pixel 6 Pro with repro steps in b/400163083#comment18
Flag: EXEMPT bugfix
(cherry picked from https://googleplex-android-review.googlesource.com/q/commit:a1d1e800712a418709cebb8faa9535ed95292b01)
Cherrypick-From: https://googleplex-android-review.googlesource.com/q/commit:3320587667e1c2d92eed83b1a0177bbc23843374
Merged-In: I3f525b44f89589d521b3d3408a976b11b1cfe064
Change-Id: I3f525b44f89589d521b3d3408a976b11b1cfe064
---
 core/java/android/content/res/Resources.java  | 212 +++++++++---------
 .../android/content/res/ResourcesImpl.java    |  99 +++++---
 .../ConfigurationBoundResourceCacheTest.java  |  12 +-
 3 files changed, 183 insertions(+), 140 deletions(-)

diff --git a/core/java/android/content/res/Resources.java b/core/java/android/content/res/Resources.java
index 92f8bb4e005e..394fa01aeb2a 100644
--- a/core/java/android/content/res/Resources.java
+++ b/core/java/android/content/res/Resources.java
@@ -574,7 +574,7 @@ public class Resources {
      * @param id The desired resource identifier, as generated by the aapt
      *           tool. This integer encodes the package, type, and resource
      *           entry. The value 0 is an invalid identifier.
-     *           
+     *
      * @param formatArgs The format arguments that will be used for substitution.
      *
      * @throws NotFoundException Throws NotFoundException if the given ID does not exist.
@@ -651,11 +651,11 @@ public class Resources {
      * Return the string value associated with a particular resource ID.  The
      * returned object will be a String if this is a plain string; it will be
      * some other type of CharSequence if it is styled.
-     * 
+     *
      * @param id The desired resource identifier, as generated by the aapt
      *           tool. This integer encodes the package, type, and resource
      *           entry. The value 0 is an invalid identifier.
-     * 
+     *
      * @param def The default CharSequence to return.
      *
      * @return CharSequence The string data associated with the resource, plus
@@ -747,25 +747,25 @@ public class Resources {
         if (len < 0) {
             throw new NotFoundException("Array resource ID #0x" + Integer.toHexString(id));
         }
-        
+
         TypedArray array = TypedArray.obtain(this, len);
         array.mLength = impl.getAssets().getResourceArray(id, array.mData);
         array.mIndices[0] = 0;
-        
+
         return array;
     }
 
     /**
-     * Retrieve a dimensional for a particular resource ID.  Unit 
+     * Retrieve a dimensional for a particular resource ID.  Unit
      * conversions are based on the current {@link DisplayMetrics} associated
      * with the resources.
-     * 
+     *
      * @param id The desired resource identifier, as generated by the aapt
      *           tool. This integer encodes the package, type, and resource
      *           entry. The value 0 is an invalid identifier.
-     * 
+     *
      * @return Resource dimension value multiplied by the appropriate metric to convert to pixels.
-     * 
+     *
      * @throws NotFoundException Throws NotFoundException if the given ID does not exist.
      *
      * @see #getDimensionPixelOffset
@@ -792,14 +792,14 @@ public class Resources {
      * {@link #getDimension}, except the returned value is converted to
      * integer pixels for you.  An offset conversion involves simply
      * truncating the base value to an integer.
-     * 
+     *
      * @param id The desired resource identifier, as generated by the aapt
      *           tool. This integer encodes the package, type, and resource
      *           entry. The value 0 is an invalid identifier.
-     * 
-     * @return Resource dimension value multiplied by the appropriate 
+     *
+     * @return Resource dimension value multiplied by the appropriate
      * metric and truncated to integer pixels.
-     * 
+     *
      * @throws NotFoundException Throws NotFoundException if the given ID does not exist.
      *
      * @see #getDimension
@@ -828,14 +828,14 @@ public class Resources {
      * integer pixels for use as a size.  A size conversion involves
      * rounding the base value, and ensuring that a non-zero base value
      * is at least one pixel in size.
-     * 
+     *
      * @param id The desired resource identifier, as generated by the aapt
      *           tool. This integer encodes the package, type, and resource
      *           entry. The value 0 is an invalid identifier.
-     * 
-     * @return Resource dimension value multiplied by the appropriate 
+     *
+     * @return Resource dimension value multiplied by the appropriate
      * metric and truncated to integer pixels.
-     *  
+     *
      * @throws NotFoundException Throws NotFoundException if the given ID does not exist.
      *
      * @see #getDimension
@@ -858,19 +858,19 @@ public class Resources {
 
     /**
      * Retrieve a fractional unit for a particular resource ID.
-     * 
+     *
      * @param id The desired resource identifier, as generated by the aapt
      *           tool. This integer encodes the package, type, and resource
      *           entry. The value 0 is an invalid identifier.
-     * @param base The base value of this fraction.  In other words, a 
+     * @param base The base value of this fraction.  In other words, a
      *             standard fraction is multiplied by this value.
-     * @param pbase The parent base value of this fraction.  In other 
+     * @param pbase The parent base value of this fraction.  In other
      *             words, a parent fraction (nn%p) is multiplied by this
      *             value.
-     * 
-     * @return Attribute fractional value multiplied by the appropriate 
+     *
+     * @return Attribute fractional value multiplied by the appropriate
      * base value.
-     *  
+     *
      * @throws NotFoundException Throws NotFoundException if the given ID does not exist.
      */
     public float getFraction(@FractionRes int id, int base, int pbase) {
@@ -886,7 +886,7 @@ public class Resources {
             releaseTempTypedValue(value);
         }
     }
-    
+
     /**
      * Return a drawable object associated with a particular resource ID.
      * Various types of objects will be returned depending on the underlying
@@ -1263,19 +1263,19 @@ public class Resources {
      * description for the given resource ID.  This parser has limited
      * functionality -- in particular, you can't change its input, and only
      * the high-level events are available.
-     * 
+     *
      * <p>This function is really a simple wrapper for calling
      * {@link #getXml} with a layout resource.
-     * 
+     *
      * @param id The desired resource identifier, as generated by the aapt
      *           tool. This integer encodes the package, type, and resource
      *           entry. The value 0 is an invalid identifier.
      *
      * @throws NotFoundException Throws NotFoundException if the given ID does not exist.
-     * 
+     *
      * @return A new parser object through which you can read
      *         the XML data.
-     *         
+     *
      * @see #getXml
      */
     @NonNull
@@ -1288,19 +1288,19 @@ public class Resources {
      * description for the given resource ID.  This parser has limited
      * functionality -- in particular, you can't change its input, and only
      * the high-level events are available.
-     * 
+     *
      * <p>This function is really a simple wrapper for calling
      * {@link #getXml} with an animation resource.
-     * 
+     *
      * @param id The desired resource identifier, as generated by the aapt
      *           tool. This integer encodes the package, type, and resource
      *           entry. The value 0 is an invalid identifier.
      *
      * @throws NotFoundException Throws NotFoundException if the given ID does not exist.
-     * 
+     *
      * @return A new parser object through which you can read
      *         the XML data.
-     *         
+     *
      * @see #getXml
      */
     @NonNull
@@ -1311,22 +1311,22 @@ public class Resources {
     /**
      * Return an XmlResourceParser through which you can read a generic XML
      * resource for the given resource ID.
-     * 
+     *
      * <p>The XmlPullParser implementation returned here has some limited
      * functionality.  In particular, you can't change its input, and only
      * high-level parsing events are available (since the document was
      * pre-parsed for you at build time, which involved merging text and
      * stripping comments).
-     * 
+     *
      * @param id The desired resource identifier, as generated by the aapt
      *           tool. This integer encodes the package, type, and resource
      *           entry. The value 0 is an invalid identifier.
      *
      * @throws NotFoundException Throws NotFoundException if the given ID does not exist.
-     * 
+     *
      * @return A new parser object through which you can read
      *         the XML data.
-     *         
+     *
      * @see android.util.AttributeSet
      */
     @NonNull
@@ -1339,9 +1339,9 @@ public class Resources {
      * with resources whose value is the name of an asset files -- that is, it can be
      * used to open drawable, sound, and raw resources; it will fail on string
      * and color resources.
-     * 
+     *
      * @param id The resource identifier to open, as generated by the aapt tool.
-     * 
+     *
      * @return InputStream Access to the resource data.
      *
      * @throws NotFoundException Throws NotFoundException if the given ID does not exist.
@@ -1414,20 +1414,20 @@ public class Resources {
      * with resources whose value is the name of an asset files -- that is, it can be
      * used to open drawable, sound, and raw resources; it will fail on string
      * and color resources.
-     * 
+     *
      * <p>This function only works for resources that are stored in the package
      * as uncompressed data, which typically includes things like mp3 files
      * and png images.
-     * 
+     *
      * @param id The resource identifier to open, as generated by the aapt tool.
-     * 
+     *
      * @return AssetFileDescriptor A new file descriptor you can use to read
      * the resource.  This includes the file descriptor itself, as well as the
      * offset and length of data where the resource appears in the file.  A
      * null is returned if the file exists but is compressed.
      *
      * @throws NotFoundException Throws NotFoundException if the given ID does not exist.
-     * 
+     *
      */
     public AssetFileDescriptor openRawResourceFd(@RawRes int id)
             throws NotFoundException {
@@ -1441,7 +1441,7 @@ public class Resources {
 
     /**
      * Return the raw data associated with a particular resource ID.
-     * 
+     *
      * @param id The desired resource identifier, as generated by the aapt
      *           tool. This integer encodes the package, type, and resource
      *           entry. The value 0 is an invalid identifier.
@@ -1461,7 +1461,7 @@ public class Resources {
 
     /**
      * Get the raw value associated with a resource with associated density.
-     * 
+     *
      * @param id resource identifier
      * @param density density in DPI
      * @param resolveRefs If true, a resource that is a reference to another
@@ -1482,10 +1482,10 @@ public class Resources {
      * See getIdentifier() for information on how names are mapped to resource
      * IDs, and getString(int) for information on how string resources are
      * retrieved.
-     * 
+     *
      * <p>Note: use of this function is discouraged.  It is much more
      * efficient to retrieve resources by identifier than by name.
-     * 
+     *
      * @param name The name of the desired resource.  This is passed to
      *             getIdentifier() with a default type of "string".
      * @param outValue Object in which to place the resource data.
@@ -1526,13 +1526,13 @@ public class Resources {
      * In other words, a Theme is a set of values for resource attributes;
      * these are used in conjunction with {@link TypedArray}
      * to resolve the final value for an attribute.
-     * 
+     *
      * <p>The Theme's attributes come into play in two ways: (1) a styled
      * attribute can explicit reference a value in the theme through the
      * "?themeAttribute" syntax; (2) if no value has been defined for a
      * particular styled attribute, as a last resort we will try to find that
      * attribute's value in the Theme.
-     * 
+     *
      * <p>You will normally use the {@link #obtainStyledAttributes} APIs to
      * retrieve XML attributes with style and theme information applied.
      */
@@ -1562,13 +1562,13 @@ public class Resources {
          * Place new attribute values into the theme.  The style resource
          * specified by <var>resid</var> will be retrieved from this Theme's
          * resources, its values placed into the Theme object.
-         * 
+         *
          * <p>The semantics of this function depends on the <var>force</var>
          * argument:  If false, only values that are not already defined in
          * the theme will be copied from the system resource; otherwise, if
          * any of the style's attributes are already defined in the theme, the
          * current values in the theme will be overwritten.
-         * 
+         *
          * @param resId The resource ID of a style resource from which to
          *              obtain attribute values.
          * @param force If true, values in the style resource will always be
@@ -1587,7 +1587,7 @@ public class Resources {
          * Resources object, they will be identical after this function
          * returns.  If they are from different Resources, only the resources
          * they have in common will be set in this theme.
-         * 
+         *
          * @param other The existing Theme to copy from.
          */
         public void setTo(Theme other) {
@@ -1601,19 +1601,19 @@ public class Resources {
         /**
          * Return a TypedArray holding the values defined by
          * <var>Theme</var> which are listed in <var>attrs</var>.
-         * 
+         *
          * <p>Be sure to call {@link TypedArray#recycle() TypedArray.recycle()} when you are done
          * with the array.
-         * 
+         *
          * @param attrs The desired attributes. These attribute IDs must be sorted in ascending
          *              order.
          *
          * @throws NotFoundException Throws NotFoundException if the given ID does not exist.
-         * 
+         *
          * @return Returns a TypedArray holding an array of the attribute values.
          * Be sure to call {@link TypedArray#recycle() TypedArray.recycle()}
          * when done with it.
-         * 
+         *
          * @see Resources#obtainAttributes
          * @see #obtainStyledAttributes(int, int[])
          * @see #obtainStyledAttributes(AttributeSet, int[], int, int)
@@ -1628,20 +1628,20 @@ public class Resources {
         /**
          * Return a TypedArray holding the values defined by the style
          * resource <var>resid</var> which are listed in <var>attrs</var>.
-         * 
+         *
          * <p>Be sure to call {@link TypedArray#recycle() TypedArray.recycle()} when you are done
          * with the array.
-         * 
+         *
          * @param resId The desired style resource.
          * @param attrs The desired attributes in the style. These attribute IDs must be sorted in
          *              ascending order.
-         * 
+         *
          * @throws NotFoundException Throws NotFoundException if the given ID does not exist.
-         * 
+         *
          * @return Returns a TypedArray holding an array of the attribute values.
          * Be sure to call {@link TypedArray#recycle() TypedArray.recycle()}
          * when done with it.
-         * 
+         *
          * @see Resources#obtainAttributes
          * @see #obtainStyledAttributes(int[])
          * @see #obtainStyledAttributes(AttributeSet, int[], int, int)
@@ -1661,13 +1661,13 @@ public class Resources {
          * that are listed in <var>attrs</var>.  In addition, if the given
          * AttributeSet specifies a style class (through the "style" attribute),
          * that style will be applied on top of the base attributes it defines.
-         * 
+         *
          * <p>Be sure to call {@link TypedArray#recycle() TypedArray.recycle()} when you are done
          * with the array.
-         * 
+         *
          * <p>When determining the final value of a particular attribute, there
          * are four inputs that come into play:</p>
-         * 
+         *
          * <ol>
          *     <li> Any attribute values in the given AttributeSet.
          *     <li> The style resource specified in the AttributeSet (named
@@ -1676,14 +1676,14 @@ public class Resources {
          *     <var>defStyleRes</var>
          *     <li> The base values in this theme.
          * </ol>
-         * 
+         *
          * <p>Each of these inputs is considered in-order, with the first listed
          * taking precedence over the following ones.  In other words, if in the
          * AttributeSet you have supplied <code>&lt;Button
          * textColor="#ff000000"&gt;</code>, then the button's text will
          * <em>always</em> be black, regardless of what is specified in any of
          * the styles.
-         * 
+         *
          * @param set The base set of attribute values.  May be null.
          * @param attrs The desired attributes to be retrieved. These attribute IDs must be sorted
          *              in ascending order.
@@ -1695,11 +1695,11 @@ public class Resources {
          *                    supplies default values for the TypedArray,
          *                    used only if defStyleAttr is 0 or can not be found
          *                    in the theme.  Can be 0 to not look for defaults.
-         * 
+         *
          * @return Returns a TypedArray holding an array of the attribute values.
          * Be sure to call {@link TypedArray#recycle() TypedArray.recycle()}
          * when done with it.
-         * 
+         *
          * @see Resources#obtainAttributes
          * @see #obtainStyledAttributes(int[])
          * @see #obtainStyledAttributes(int, int[])
@@ -1741,7 +1741,7 @@ public class Resources {
          * Retrieve the value of an attribute in the Theme.  The contents of
          * <var>outValue</var> are ultimately filled in by
          * {@link Resources#getValue}.
-         * 
+         *
          * @param resid The resource identifier of the desired theme
          *              attribute.
          * @param outValue Filled in with the ultimate resource value supplied
@@ -1750,7 +1750,7 @@ public class Resources {
          *                    false, <var>outValue</var> may be a
          *                    TYPE_REFERENCE.  In either case, it will never
          *                    be a TYPE_ATTRIBUTE.
-         * 
+         *
          * @return boolean Returns true if the attribute was found and
          *         <var>outValue</var> is valid, else false.
          */
@@ -1813,7 +1813,7 @@ public class Resources {
 
         /**
          * Print contents of this theme out to the log.  For debugging only.
-         * 
+         *
          * @param priority The log priority to use.
          * @param tag The log tag to use.
          * @param prefix Text to prefix each line printed.
@@ -2163,7 +2163,7 @@ public class Resources {
      * @return Returns a TypedArray holding an array of the attribute values.
      * Be sure to call {@link TypedArray#recycle() TypedArray.recycle()}
      * when done with it.
-     * 
+     *
      * @see Theme#obtainStyledAttributes(AttributeSet, int[], int, int)
      */
     public TypedArray obtainAttributes(AttributeSet set, @StyleableRes int[] attrs) {
@@ -2252,10 +2252,10 @@ public class Resources {
     }
 
     /**
-     * Return the current configuration that is in effect for this resource 
+     * Return the current configuration that is in effect for this resource
      * object.  The returned object should be treated as read-only.
-     * 
-     * @return The resource's current configuration. 
+     *
+     * @return The resource's current configuration.
      */
     public Configuration getConfiguration() {
         return mResourcesImpl.getConfiguration();
@@ -2274,7 +2274,7 @@ public class Resources {
     /**
      * Return the compatibility mode information for the application.
      * The returned object should be treated as read-only.
-     * 
+     *
      * @return compatibility info.
      * @hide
      */
@@ -2294,16 +2294,16 @@ public class Resources {
             mResourcesImpl.updateConfiguration(null, null, ci);
         }
     }
-    
+
     /**
      * Return a resource identifier for the given resource name.  A fully
      * qualified resource name is of the form "package:type/entry".  The first
      * two components (package and type) are optional if defType and
      * defPackage, respectively, are specified here.
-     * 
+     *
      * <p>Note: use of this function is discouraged.  It is much more
      * efficient to retrieve resources by identifier than by name.
-     * 
+     *
      * @param name The name of the desired resource.
      * @param defType Optional default resource type to find, if "type/" is
      *                not included in the name.  Can be null to require an
@@ -2311,7 +2311,7 @@ public class Resources {
      * @param defPackage Optional default package to find, if "package:" is
      *                   not included in the name.  Can be null to require an
      *                   explicit package.
-     * 
+     *
      * @return int The associated resource identifier.  Returns 0 if no such
      *         resource was found.  (0 is not a valid resource ID.)
      */
@@ -2336,13 +2336,13 @@ public class Resources {
     /**
      * Return the full name for a given resource identifier.  This name is
      * a single string of the form "package:type/entry".
-     * 
+     *
      * @param resid The resource identifier whose name is to be retrieved.
-     * 
+     *
      * @return A string holding the name of the resource.
-     * 
+     *
      * @throws NotFoundException Throws NotFoundException if the given ID does not exist.
-     * 
+     *
      * @see #getResourcePackageName
      * @see #getResourceTypeName
      * @see #getResourceEntryName
@@ -2350,33 +2350,33 @@ public class Resources {
     public String getResourceName(@AnyRes int resid) throws NotFoundException {
         return mResourcesImpl.getResourceName(resid);
     }
-    
+
     /**
      * Return the package name for a given resource identifier.
-     * 
+     *
      * @param resid The resource identifier whose package name is to be
      * retrieved.
-     * 
+     *
      * @return A string holding the package name of the resource.
-     * 
+     *
      * @throws NotFoundException Throws NotFoundException if the given ID does not exist.
-     * 
+     *
      * @see #getResourceName
      */
     public String getResourcePackageName(@AnyRes int resid) throws NotFoundException {
         return mResourcesImpl.getResourcePackageName(resid);
     }
-    
+
     /**
      * Return the type name for a given resource identifier.
-     * 
+     *
      * @param resid The resource identifier whose type name is to be
      * retrieved.
-     * 
+     *
      * @return A string holding the type name of the resource.
-     * 
+     *
      * @throws NotFoundException Throws NotFoundException if the given ID does not exist.
-     * 
+     *
      * @see #getResourceName
      */
     public String getResourceTypeName(@AnyRes int resid) throws NotFoundException {
@@ -2412,13 +2412,13 @@ public class Resources {
     public String getLastResourceResolution() throws NotFoundException {
         return mResourcesImpl.getLastResourceResolution();
     }
-    
+
     /**
      * Parse a series of {@link android.R.styleable#Extra &lt;extra&gt;} tags from
      * an XML file.  You call this when you are at the parent tag of the
      * extra tags, and it will return once all of the child tags have been parsed.
      * This will call {@link #parseBundleExtra} for each extra tag encountered.
-     * 
+     *
      * @param parser The parser from which to retrieve the extras.
      * @param outBundle A Bundle in which to place all parsed extras.
      * @throws XmlPullParserException
@@ -2433,7 +2433,7 @@ public class Resources {
             if (type == XmlPullParser.END_TAG || type == XmlPullParser.TEXT) {
                 continue;
             }
-            
+
             String nodeName = parser.getName();
             if (nodeName.equals("extra")) {
                 parseBundleExtra("extra", parser, outBundle);
@@ -2442,9 +2442,9 @@ public class Resources {
             } else {
                 XmlUtils.skipCurrentTag(parser);
             }
-        }        
+        }
     }
-    
+
     /**
      * Parse a name/value pair out of an XML tag holding that data.  The
      * AttributeSet must be holding the data defined by
@@ -2459,7 +2459,7 @@ public class Resources {
      * <li> {@link TypedValue#TYPE_FLOAT}:
      * {@link Bundle#putCharSequence Bundle.putFloat()}
      * </ul>
-     * 
+     *
      * @param tagName The name of the tag these attributes come from; this is
      * only used for reporting error messages.
      * @param attrs The attributes from which to retrieve the name/value pair.
@@ -2508,7 +2508,7 @@ public class Resources {
 
         sa.recycle();
     }
-    
+
     /**
      * Retrieve underlying AssetManager storage for these resources.
      */
@@ -2533,7 +2533,7 @@ public class Resources {
     public final void startPreloading() {
         mResourcesImpl.startPreloading();
     }
-    
+
     /**
      * Called by zygote when it is done preloading resources, to change back
      * to normal Resources operation.
@@ -2614,12 +2614,16 @@ public class Resources {
     }
 
     /**
-     * Called by ConfigurationBoundResourceCacheTest.
+     * Called by ConfigurationBoundResourceCacheTest. Applies the new configuration, returning a
+     * bitmask of the changes between the old and new configurations.
+     *
+     * @param config the new configuration
+     * @return bitmask of config changes
      * @hide
      */
     @VisibleForTesting
-    public int calcConfigChanges(Configuration config) {
-        return mResourcesImpl.calcConfigChanges(config);
+    public int applyConfigChanges(Configuration config) {
+        return mResourcesImpl.applyConfigChanges(config);
     }
 
     /**
diff --git a/core/java/android/content/res/ResourcesImpl.java b/core/java/android/content/res/ResourcesImpl.java
index 6cbad2f0909b..9dbb564e5262 100644
--- a/core/java/android/content/res/ResourcesImpl.java
+++ b/core/java/android/content/res/ResourcesImpl.java
@@ -467,35 +467,56 @@ public class ResourcesImpl {
                 // the framework.
                 mDisplayAdjustments.getCompatibilityInfo().applyToDisplayMetrics(mMetrics);
 
-                final @Config int configChanges = calcConfigChanges(config);
+                final @Config int configChanges = applyConfigChanges(config);
 
                 // If even after the update there are no Locales set, grab the default locales.
-                LocaleList locales = mConfiguration.getLocales();
-                if (locales.isEmpty()) {
-                    locales = LocaleList.getDefault();
-                    mConfiguration.setLocales(locales);
+                LocaleList configLocales = mConfiguration.getLocales();
+                if (configLocales.isEmpty()) {
+                    configLocales = LocaleList.getDefault();
+                    mConfiguration.setLocales(configLocales);
                 }
 
+                // Note: `locales` should be a list in the following order:
+                // 1. App-specific locales (typically at most one, set by the user in settings)
+                // 2. One or more system locales
+
                 String[] selectedLocales = null;
                 String defaultLocale = null;
                 LocaleConfig lc = ResourcesManager.getInstance().getLocaleConfig();
                 if ((configChanges & ActivityInfo.CONFIG_LOCALE) != 0) {
-                    if (locales.size() > 1) {
+                    if (configLocales.size() > 1) {
+                        // There is more than one locale in the configuration. We need to update
+                        // the locales in the configuration based on what is supported by the app
+                        // so that we will use the same locales on other non-locale config changes.
+
+                        Locale[] intersection = null;
                         if (Flags.defaultLocale() && (lc.getDefaultLocale() != null)) {
-                            Locale[] intersection =
-                                    locales.getIntersection(lc.getSupportedLocales());
-                            mConfiguration.setLocales(new LocaleList(intersection));
-                            selectedLocales = new String[intersection.length];
-                            for (int i = 0; i < intersection.length; i++) {
-                                selectedLocales[i] =
-                                        adjustLanguageTag(intersection[i].toLanguageTag());
+                            // Note: getIntersection() returns a list of locales in the same order
+                            // currently in `configLocales`, which is desired (i.e., app-specific
+                            // locales first before system locales).
+                            defaultLocale = adjustLanguageTag(
+                                    lc.getDefaultLocale().toLanguageTag());
+                            intersection =
+                                    configLocales.getIntersection(
+                                            lc.getSupportedLocales());
+                            if (intersection.length > 0) {
+                                mConfiguration.setLocales(new LocaleList(intersection));
+                                selectedLocales = new String[intersection.length];
+                                for (int i = 0; i < intersection.length; i++) {
+                                    selectedLocales[i] =
+                                            adjustLanguageTag(intersection[i].toLanguageTag());
+                                }
+                                Slog.v(TAG, "Updating configuration, with default locale "
+                                        + defaultLocale + " and selected locales "
+                                        + Arrays.toString(selectedLocales));
                             }
-                            defaultLocale =
-                                    adjustLanguageTag(lc.getDefaultLocale().toLanguageTag());
-                            Slog.v(TAG, "Updating configuration, with default locale "
-                                    + defaultLocale + " and selected locales "
-                                    + Arrays.toString(selectedLocales));
-                        } else {
+                        }
+
+                        if (intersection == null || intersection.length == 0) {
+                            // This is the fallback behavior when multi-locale is not enabled,
+                            // or when there was no intersection between the app's supported
+                            // locales and the locales in the configuration.
+
                             String[] availableLocales;
                             // The LocaleList has changed. We must query the AssetManager's
                             // available Locales and figure out the best matching Locale in the new
@@ -510,30 +531,45 @@ public class ResourcesImpl {
                             }
 
                             if (availableLocales != null) {
-                                final Locale bestLocale = locales.getFirstMatchWithEnglishSupported(
-                                        availableLocales);
+                                final Locale bestLocale =
+                                        configLocales.getFirstMatchWithEnglishSupported(
+                                                availableLocales);
                                 if (bestLocale != null) {
                                     selectedLocales = new String[]{
                                             adjustLanguageTag(bestLocale.toLanguageTag())};
-                                    if (!bestLocale.equals(locales.get(0))) {
+                                    if (!bestLocale.equals(configLocales.get(0))) {
+                                        // There was an locale available in the app that matched.
+                                        // Update the configuration so that it goes first.
                                         mConfiguration.setLocales(
-                                                new LocaleList(bestLocale, locales));
+                                                new LocaleList(bestLocale, configLocales));
+                                        Slog.v(TAG, "Updating configuration with selected locales "
+                                                + Arrays.toString(selectedLocales));
                                     }
                                 }
                             }
                         }
                     }
                 }
-                if (selectedLocales == null) {
+
+                if (selectedLocales == null || selectedLocales.length == 0) {
+                    // Either:
+                    // 1. The locales were not changed in the configuration, in which case we
+                    //    use whatever was last set in the configuration.
+                    // 2. Locales changed, but there was only one locale in the configuration, in
+                    //    which case we simply use that locale.
+                    // 3. Locales changed, and there were multiple locales in the configuration,
+                    //    but the app does not support any of them, so we use whatever locales are
+                    //    currently in the configuration.
                     if (Flags.defaultLocale() && (lc.getDefaultLocale() != null)) {
-                        selectedLocales = new String[locales.size()];
-                        for (int i = 0; i < locales.size(); i++) {
-                            selectedLocales[i] = adjustLanguageTag(locales.get(i).toLanguageTag());
+                        selectedLocales = new String[configLocales.size()];
+                        for (int i = 0; i < configLocales.size(); i++) {
+                            selectedLocales[i] =
+                                    adjustLanguageTag(configLocales.get(i).toLanguageTag());
                         }
                         defaultLocale = adjustLanguageTag(lc.getDefaultLocale().toLanguageTag());
                     } else {
                         selectedLocales = new String[]{
-                                adjustLanguageTag(locales.get(0).toLanguageTag())};
+                                adjustLanguageTag(configLocales.get(0).toLanguageTag())};
                     }
                 }
 
@@ -613,7 +649,7 @@ public class ResourcesImpl {
      * @param config the new configuration
      * @return bitmask of config changes
      */
-    public @Config int calcConfigChanges(@Nullable Configuration config) {
+    public @Config int applyConfigChanges(@Nullable Configuration config) {
         if (config == null) {
             // If there is no configuration, assume all flags have changed.
             return 0xFFFFFFFF;
@@ -859,7 +895,10 @@ public class ResourcesImpl {
                 }
             } else {
                 if (verifyPreloadConfig(
-                        changingConfigs, ActivityInfo.CONFIG_LAYOUT_DIRECTION, value.resourceId, "drawable")) {
+                        changingConfigs,
+                        ActivityInfo.CONFIG_LAYOUT_DIRECTION,
+                        value.resourceId,
+                        "drawable")) {
                     if ((changingConfigs & ActivityInfo.CONFIG_LAYOUT_DIRECTION) == 0) {
                         // If this resource does not vary based on layout direction,
                         // we can put it in all of the preload maps.
diff --git a/core/tests/coretests/src/android/content/res/ConfigurationBoundResourceCacheTest.java b/core/tests/coretests/src/android/content/res/ConfigurationBoundResourceCacheTest.java
index 68882eb39e37..effbd40cc5a4 100644
--- a/core/tests/coretests/src/android/content/res/ConfigurationBoundResourceCacheTest.java
+++ b/core/tests/coretests/src/android/content/res/ConfigurationBoundResourceCacheTest.java
@@ -111,7 +111,7 @@ public class ConfigurationBoundResourceCacheTest {
         newCnf.orientation = cfg.orientation == Configuration.ORIENTATION_LANDSCAPE ?
                 Configuration.ORIENTATION_PORTRAIT
                 : Configuration.ORIENTATION_LANDSCAPE;
-        int changes = calcConfigChanges(res, newCnf);
+        int changes = applyConfigChanges(res, newCnf);
         assertEquals(staticDim, mCache.getInstance(key, res, mContext.getTheme()));
         mCache.onConfigurationChange(changes);
         assertEquals(staticDim, mCache.getInstance(key, res, mContext.getTheme()));
@@ -134,7 +134,7 @@ public class ConfigurationBoundResourceCacheTest {
         newCnf.orientation = cfg.orientation == Configuration.ORIENTATION_LANDSCAPE ?
                 Configuration.ORIENTATION_PORTRAIT
                 : Configuration.ORIENTATION_LANDSCAPE;
-        int changes = calcConfigChanges(res, newCnf);
+        int changes = applyConfigChanges(res, newCnf);
         assertEquals(changingDim,
                 mCache.getInstance(key, res, mContext.getTheme()));
         mCache.onConfigurationChange(changes);
@@ -162,7 +162,7 @@ public class ConfigurationBoundResourceCacheTest {
         newCnf.orientation = cfg.orientation == Configuration.ORIENTATION_LANDSCAPE ?
                 Configuration.ORIENTATION_PORTRAIT
                 : Configuration.ORIENTATION_LANDSCAPE;
-        int changes = calcConfigChanges(res, newCnf);
+        int changes = applyConfigChanges(res, newCnf);
         assertEquals(staticDim, mCache.getInstance(R.dimen.resource_cache_test_generic, res,
                 mContext.getTheme()));
         assertEquals(changingDim,
@@ -205,7 +205,7 @@ public class ConfigurationBoundResourceCacheTest {
         newCnf.orientation = cfg.orientation == Configuration.ORIENTATION_LANDSCAPE ?
                 Configuration.ORIENTATION_PORTRAIT
                 : Configuration.ORIENTATION_LANDSCAPE;
-        int changes = calcConfigChanges(res, newCnf);
+        int changes = applyConfigChanges(res, newCnf);
         for (int i = 0; i < 2; i++) {
             final Resources.Theme theme = i == 0 ? mContext.getTheme() : null;
             assertEquals(staticDim,
@@ -224,8 +224,8 @@ public class ConfigurationBoundResourceCacheTest {
         }
     }
 
-    private static int calcConfigChanges(Resources resources, Configuration configuration) {
-        return resources.calcConfigChanges(configuration);
+    private static int applyConfigChanges(Resources resources, Configuration configuration) {
+        return resources.applyConfigChanges(configuration);
     }
 
     static class DummyFloatConstantState extends ConstantState<Float> {
-- 
2.43.0

